# NumPy Final Project
## Student Exam Analysis System

Imagine a school gives marks in 3 subjects:

Math\
Science\
English

And we have marks for 5 students.

### Step 1: Create Dataset

import numpy as np

marks = np.array([
    [80, 75, 90],
    [60, 70, 65],
    [95, 88, 92],
    [50, 55, 60],
    [85, 80, 78]
])

print(marks)

Understand the Data

_____________________________________
|Student |  Math | Science| English |
|:------:|:-----:|:------:|:-------:|
|1       |   80  |   75   |    90   |
|2       |   60  |   70   |    65   |
|3       |   95  |   88   |    92   |
|4       |   50  |   55   |    60   |
|5       |   85  |   80   |    78   |


Shape:

In [2]:
print(marks.shape)

(5, 3)


Meaning:

5 Students \
3 Subjects

### Step 2: Basic Statistics

In [3]:
# Total Students

print(marks.shape[0])

5


In [4]:
#Total Subjects

print(marks.shape[1])

3


In [5]:
# Overall Average

print(marks.mean())

74.86666666666666


### Step 3: Subject-Wise Average

In [6]:
print(marks.mean(axis=0))

[74.  73.6 77. ]


Meaning:

Math Average    = 74

Science Average = 73.6

English Average = 77

Math Column

80 \
60 \
95 \
50 \
85

Average = 74

### Step 4: Student-Wise Average

In [9]:
print(marks.mean(axis=1))

[81.66666667 65.         91.66666667 55.         81.        ]


Meaning:

Student 1 Average \
Student 2 Average \
Student 3 Average \
Student 4 Average \
Student 5 Average 

### Step 5: Highest Marks

In [10]:
print(marks.max())

95


### Step 6: Lowest Marks

In [11]:
print(marks.min())

50


### Step 7: Find Top Student

First calculate student averages.

In [12]:
student_avg = marks.mean(axis=1)

print(student_avg)

[81.66666667 65.         91.66666667 55.         81.        ]


Find highest average.

In [13]:
top_student = student_avg.argmax()

print(top_student)

2


Remember: Python starts counting from 0

So: Student 3 is topper
    
### Step 8: Passing Students

Suppose: Pass = Average ≥ 60

Calculate averages:

In [17]:
student_avg = marks.mean(axis=1)

# Get passing students:

passed = student_avg >= 60

print(passed)

[ True  True  True False  True]


Only passing averages:

In [18]:
print(student_avg[student_avg >= 60])

[81.66666667 65.         91.66666667 81.        ]


### Step 9: Add Bonus Marks

Teacher gives: +5 Marks to Everyone

In [19]:
bonus_marks = marks + 5

print(bonus_marks)

[[ 85  80  95]
 [ 65  75  70]
 [100  93  97]
 [ 55  60  65]
 [ 90  85  83]]


### Step 10: Flatten Dataset

Convert to one long array.

In [20]:
print(marks.flatten())

[80 75 90 60 70 65 95 88 92 50 55 60 85 80 78]


### Step 11: Sort All Marks

In [22]:
sorted_marks = np.sort(marks.flatten())

print(sorted_marks)

[50 55 60 60 65 70 75 78 80 80 85 88 90 92 95]


## Full Project Code

In [26]:
import numpy as np

marks = np.array([
    [80,75,90],
    [60,70,65],
    [95,88,92],
    [50,55,60],
    [85,80,78]
])

print("Dataset:")
print(marks)

print("\nShape:")
print(marks.shape)

print("\nOverall Average:")
print(marks.mean())

print("\nSubject Averages:")
print(marks.mean(axis=0))

student_avg = marks.mean(axis=1)

print("\nStudent Averages:")
print(student_avg)

print("\nHighest Mark:")
print(marks.max())

print("\nLowest Mark:")
print(marks.min())

top_student = student_avg.argmax()

print("\nTop Student Index:")
print(top_student)

print("\nStudents Passed:")
print(student_avg[student_avg >= 60])

print("\nBonus Marks:")
print(marks + 5)

print("\nFlatten:")
print(marks.flatten())

print("\nSorted Marks:")
print(np.sort(marks.flatten()))

Dataset:
[[80 75 90]
 [60 70 65]
 [95 88 92]
 [50 55 60]
 [85 80 78]]

Shape:
(5, 3)

Overall Average:
74.86666666666666

Subject Averages:
[74.  73.6 77. ]

Student Averages:
[81.66666667 65.         91.66666667 55.         81.        ]

Highest Mark:
95

Lowest Mark:
50

Top Student Index:
2

Students Passed:
[81.66666667 65.         91.66666667 81.        ]

Bonus Marks:
[[ 85  80  95]
 [ 65  75  70]
 [100  93  97]
 [ 55  60  65]
 [ 90  85  83]]

Flatten:
[80 75 90 60 70 65 95 88 92 50 55 60 85 80 78]

Sorted Marks:
[50 55 60 60 65 70 75 78 80 80 85 88 90 92 95]


What we Learned in This Project

* Arrays
* 2D Arrays
* Shape
* Indexing
* Slicing
* Mean
* Max
* Min
* Boolean Filtering
* Argmax
* Flatten
* Sort
* Array Operations

This is a real beginner-level data analysis workflow.

### Extra Challenge

In [28]:
print("Median:", np.median(marks))
print("Std:", np.std(marks))

Median: 78.0
Std: 13.764043333588022
